weeks/week05/modules.md

```diff
@@ -1,3 +1,150 @@
+
+Using the module name you can access the functions:
+
+### Example:
+```python
+import time
+time.sleep(30)  # Access sleep() from the standard module. Only the module name is added to the current namespace.
+```
+
+Instead of importing the whole module, you can import only some names:
+
+### Syntax:
+```python
+from [module name] import [name]
+```
+
+### Example:
+```python
+from time import sleep
+sleep(30)  # sleep name is imported from a module directly into the importing module’s namespace. 
+```
+
+It is also possible to import an object but add it with an alternate name:
+
+### Syntax:
+```python
+from [module name] import [name] as [your module alias]
+```
+
+### Example:
+```python
+from time import sleep as my_local_sleep
+my_local_sleep(1) 
+```
+
+There is a variant to import all names that a module defines:
+
+### Syntax:
+```python
+from [module name] import *
+```
+
+This imports all names except those beginning with an underscore `_`. Try to avoid using this facility since it introduces an unknown set of names and might overwrite an existing name as Python will call the method from the last import.
+
+## 1.3 Searching Modules
+
+When the interpreter executes the `import` statement,  it searches for the name of the module in the following sources:
+
+- The directory from which the input script was run or the current directory if the interpreter is being run interactively
+- The list of directories contained in the `PYTHONPATH` environment variable, if it is set. (The format for PYTHONPATH is OS-dependent but should mimic the PATH environment variable.)
+- An installation-dependent list of directories configured at the time Python is installed
+
+### Example:
+```python
+import sys
+print(sys.path) # Output will be the resulting search path 
+```
+
+Once a module has been imported, you can determine the location where it was found with the module’s `__file__` attribute:
+
+### Example:
+```python
+import re
+re.__file__ # Output: re.py file location 
+```
+
+## 1.4 Executing Module
+
+Any `.py` file that contains a module is essentially also a Python script and can be executed like one. Python runs all the code before importing it.
+
+When a .py file is imported as a module, Python sets the special dunder variable `__name__` to the name of the module. However, if a file is run as a standalone script, `__name__` is set to the string `'__main__'`. This can be used to distinguish between when the file is loaded as a module and when it is run as a standalone script.
+
+### Example:
+```python
+if (__name__ == '__main__'):
+    print('Executing as standalone script')
+```
+
+## 1.5 Cyclic Imports
+
+Suppose you have the following two files: `module_a.py` and `module_b.py`. Each file tries to import functions from one another.
+
+### Example:
+```python
+from import_b import boo
```

Comment:
Add a comment just to show that this is a file, like `# module_a.py` and the other one `# module_b.py`.